Create a guide to configure Azure OpenID with Okteto (#259)
okteto/docs: src/content/self-hosted/guides/azure/index.mdx

---

> title: Use Azure as your Authentication Provider

Reviewer: ```suggestion
title: Use Azure Active Directory as your Authentication Provider
```

> description: Guide on how use Azure as the authentication provider of your Okteto Self-hosted instance

Reviewer: ```suggestion
description: How to use Azure Active Directory as the authentication provider of your Okteto Self-hosted instance
```

id: identity-provider-azure
---

import Image from '@theme/Image';
import TiersList from '@theme/TiersList';

#  Use Azure as your Identity Provider<TiersList tiers="Self-Hosted" />

This guide will walk you through the steps necessary to configure Azure as the authentication provider for your instace of Okteto.

Please refer to [Azure's official documentation](https://learn.microsoft.com/en-us/power-apps/maker/portals/configure/configure-openid-settings) on Open ID Connect for more information.

## Prerequisites

- An instance of [Okteto](https://okteto.com/docs/self-hosted/).
- [Helm](https://helm.sh/docs/intro/install/) 3.0+ installed in your local machine.
- Access to an [Azure account](https://portal.azure.com) with permissions to registrate applications in Azure Active Directory.

## Create an App Registration

> Create a new app integration, using the values below:

Reviewer: Where do they create the app integration? Can we include a link or instructions on where on the portal to do this?

> <p align="center"><Image src="/docs/self-hosted/guides/azure/images/new-app.png" alt="Create the new App Registration" /></p>

Reviewer: don't use `demo.okteto.dev` for the URL since that uses public override, that is not fully documented (we should write a guide on that). Use `okteto.example.com`, that's what we used for the Okta guide.

Then, fill in the following values on the “Register an application” window:

> 1. Name: The name of your app (e.g. Okteto)

Reviewer: don't use abbreviations if possible.
```suggestion
1. Name: The name of your application (e.g. Okteto)
```

Author: `app` is the name used by Azure

2. Accounts in this organizational directory only (Default Directory only - Single tenant)
3. Redirect URI: https://okteto.{YOUR_DOMAIN}/auth/callback

> Once the application has been created, copy the `Application (client) ID` value. This will be your `<<client-id>>`.

Reviewer: At this point we haven't said what the <<client-id>> is.  Also, why use that format? In the okta guideswe use a different format. we should use the same here. 

```suggestion
Once the application has been created, save the `Application (client) ID` value. You'll need it to configure your Okteto Instance.
```

> You'll need it to configure your Okteto Instance.

Reviewer: ```suggestion

```

<p align="center"><Image src="/docs/self-hosted/guides/azure/images/client-id.png" alt="Copy the Tenant ID" /></p>

> Also, copy the `Directory (tenant) ID` value. This will be your `<<tenant-id>>`.

Reviewer: same as above. since we haven't told them what the tenant-id is, i don't think we should mention it. and if we do, we should use the same format than in other guides.

```suggestion
Copy the `Directory (tenant) ID` value`. You'll need it to configure your Okteto Instance.
```

You'll need it to configure your Okteto Instance.

<p align="center"><Image src="/docs/self-hosted/guides/azure/images/tenant-id.png" alt="Copy the Tenant ID" /></p>

> On the left menu, click on "Certificates & secrets". Create a "New client secret" and copy the `Value`. This will be your `<<client-secret>>`.

Reviewer: ```suggestion
On the left menu, click on "Certificates & secrets". Create a "New client secret" and copy the `Value`. You'll need it to configure your Okteto Instance.
```

<p align="center"><Image src="/docs/self-hosted/guides/azure/images/secrets.png" alt="Copy the Secret ID and Value" /></p>

On the lest menu, click on "API permissions" and grant the following permissions:

<p align="center"><Image src="/docs/self-hosted/guides/azure/images/permissions.png" alt="Grant permissions" /></p>

## Configure Okteto

Once you have the Okta application ready, you'll need to update the configuration of your Okteto Self-Hosted instance. 

First, update the auth section of your Helm configuration file (e.g. `values.yaml` ) with the values you obtained on the first step of this guide:

> ```yaml
> auth:
>   openid:
>     enabled: true
>     clientId: <<client-id>>
>     clientSecret: <<client-secret>>
>     endpoints:
>       issuer: https://sts.windows.net/<<tenant-id>>/
>       authorization: https://login.microsoftonline.com/<<tenant-id>>/oauth2/v2.0/authorize
>     mapping:
>       externalIDKey: email

Reviewer: ```suggestion
```yaml
auth:
  openid:
    enabled: true
    clientId: $CLIENT_ID
    clientSecret: $CLIENT_SECRET
    endpoints:
      issuer: https://sts.windows.net/$TENANT_ID/
      authorization: https://login.microsoftonline.com/<<tenant-id>>/oauth2/v2.0/authorize
    mapping:
      externalIDKey: email
```

```

> You can also store the value of clientSecret [as a secret directly in Kubernetes](/docs/self-hosted/install/deployment/#cloud-provider-secret).

Then, [upgrade your installation](/docs/self-hosted/install/upgrade/) to apply the changes. We recommend that you upgrade to the same version that you already have to minimize the changes and help with your troubleshooting. 

Once the upgrade has been completed, go to your instance and click on the 'Log In' button to access your instance using Okta as the authentication provider.

<p align="center"><Image src="/docs/self-hosted/guides/azure/images/with-azure.png" alt="Log in with Azure" /></p>

### Customize the attributes that Okteto consumes

By default, Okteto will use the following attributes below when creating a user:

- `Name` as the display name of the user.
- `Nickname` as the username of the user.
- `Email` as the email.
- `Picture URL` for the user's profile picture.

If you are not using some of those values, or you want to customize which attributes get mapped, you can customize that by using the `mapping` field inside the `openid` key in your Helm configuration file.

```yaml
auth:
  openid:
    enabled: true

>     clientId: <<client-id>>

Reviewer: Can you use the same format for the variables that we use in the other guides? (I don't have a strong preference on which style, but it should be the same everywhere.
